### ISSUE TITLE
Retry sending messages to coralogix on intermittent network outages

### ISSUE THREAD

**Issue by rofe** [adobe/helix-log#60]
**Description**
Errors like the one below are visible in Coralogix on a daily basis (I masked some of the data I deemed to be sensitive):
```json
{
   "exception"  : {
     "$type"  :  "Error" ,
     "name"  :  "Error" ,
     "message"  :  "connect ECONNREFUSED 54.229.116.198:443" ,
     "stack"  :  "Error: connect ECONNREFUSED 54.229.116.198:443
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1104:14)" ,
     "code"  :  "ECONNREFUSED" 
  },
   "level"  :  "error" ,
   "logger"  : {
     "$type"  :  "CoralogixLogger" ,
     "level"  :  "info" ,
     "filter"  : {
       "$type"  :  "Function" 
    },
     "formatter"  : {
       "$type"  :  "Function" 
    },
     "apikey"  : {
       "$type"  :  "Secret" 
    },
     "apiurl"  :  "https://api.coralogix.com/api/v1/" ,
     "app"  :  "..." ,
     "subsystem"  :  "..." ,
     "host"  :  "..." 
  },
   "ow"  : { ... },
   "message"  :  "Encountered exception while logging!" ,
   "timestamp"  :  "2019-10-27T20:39:15.530Z" 
}
```

**To Reproduce**
Log into Coralogix and search for this type of error, or click the link in the daily error report email.

**Expected behavior**
No errors during logging.

@koraa

**Comment by rofe** [comment]
@koraa did you have a chance to look into this yet? These are currently the only errors left in Coralogix. I think it would be great to start with a clean slate, i.e. zero errors :)

**Comment by koraa** [comment]
Thanks for bumping this!

> connect ECONNREFUSED 54.229.116.198:443

This is a networking error, indicating that a connection to the given host (the coralogix server) could not be established. Usually this indicates that the uplink to coralogix is down.

Analyzing the error logs in coralogix, there are a few bursts of such errors during the last seven days. This and the fact that the error message itself could be sent (which is done after a brief pause) indicates this too.

> I think it would be great to start with a clean slate, i.e. zero errors :)

> Expected behavior
> No errors during logging.

I think that there are three categories of errors we need to consider: developer error (i.e. bugs), user error, system limitations. Category 1 – developer errors – is the only one we can really fix; the second category of errors should be displayed to a user for manual repair.

The third category of errors is a bit more complicated to deal with: These are things like bit flips in memory/cables/wifi or other kinds of electrical failure that *will* happen. We will never get these down to zero, but we can drastically reduce the error rate however by designing our systems to be resilient against such errors.

In the specific case we can probably just wait for a set amount of time 1ms, 5ms, 20ms and retry transmission. This should drastically reduce the rate of such errors. If we want to guard against longer waiting periods, we need to think about more sophisticated queue management to guard against potential DOS attacks utilizing feedback loops in our system (downtime tirggering more errors -> more to hold in memory -> oom -> more downtime).

I would like to note, that fact that we even see those errors is not a bug but an improvement in our logging infrastructure. All past logging systems where fire and forget so we simply wouldn't have seen the errors. At one such errors every 20k messages on average, these downtimes do not have a big impact on our systems eitehr…

Feel free to open a PR for retrying if you feel like it; I'll get to it at some point but I can't promise when…

**Comment by rofe** [comment]
> Feel free to open a PR for retrying

Thanks @koraa, but I think I'll leave that to the module expert ;) 

Btw, I fully agree with you that it's great we are now actually seeing _any_ errors thanks to `helix-log`, but I also believe we should fix them. So I disagree with moving this particular issue from _bug_ to _enhancement_ - after all, it's responsible for the only errors currently present in the Helix logs...

**Comment by koraa** [comment]
Thanks for your comments.

I do agree we should do something about it. I moved this from bug to enhancement because this is not an actual bug in the helix log code, it's an actual issue with the underlying infrastructure (we are experiencing intermittent network outages).

As I explained before, this is an error we cannot truly fix as we are not the ones causing it. The error needs to be fixed on the networking level, but it can (and should) be mitigated on helix log level.

Anyways, it doesn't really matter if this is a bug or an enhancement…it needs doing in either case…

**Comment by rofe** [comment]
Unless I'm missing something, I think these connection errors are still popping up in Coralogix. Reopening for clarification.

**Comment by koraa** [comment]
Hi,

I cannot find any "ECONNREFUSED" errors in the last seven days in the coralogix logs…

What I can see is 11 ETIMEDOUT errors during the last 8 days. All of them occurred since the second December and they are solitary errors (maybe one an hour). ETIMEDOUT is another error that indicates a shaky internet connection, maybe this was a slightly different outage length – the mechanism introduced with this catches this category of error too…

Given that this is another error that will occur once in a while and that this affects 0.007% of messages we sent (1 in 15k) I am not concerned. We're at a scale here (150k msgs) where statistically very rare errors will occur…

**Comment by rofe** [comment]
> What I can see is 11 ETIMEDOUT errors during the last 8 days.

You are right. Different error. 

Not sure I follow your math, though: I searched for errors in `helix-pages` during the last 24h, where it was 4 out of 14. So when excluding the INFO log level, which is the first thing you do when investigating a problem, I wouldn't call it statistically insignificant.

**Comment by koraa** [comment]
When you exclude 99.999% of the messages – you are right; suddenly the message is quite prominent ;)

The code tries to connect, retries and retries and finally gives up after a couple of MS. What specifically about that would you like to change about that behavior?

**Comment by rofe** [comment]
Maybe the time out could be increased on the retrials?

**Comment by koraa** [comment]
We could yes…there is a trade of here though…the time outs are intentionally chosen conservatively to avoid the risk of eating too much memory in outage situations…especially since we then would expect very many messages at the same time…

Of course I have no idea how much I would have to increase the time out and what risk that would introduce…I would have to run trials…

Let me ask another question; why do those eleven errors even bother you? Reducing the error rate from 0.007% to 0.0007% or even 0.00007% is probably possible, but why?

**Comment by rofe** [comment]
Probably because they appear very prominently in my inbox every morning as part of the Coralogix daily error report 😛 my point is: we are currently in the very comfortable situation of having an extremely low number of errors thrown by our services, which makes it super easy to spot "real" problems when they occur. I feel that if we let those get sprinkled with infrastructure related problems too much, we are running the risk of losing this advantage down the road...

Bottom line: I just wanted to discuss it here rather than implicitly assume it should be ignored. I don't want to belabor the point. I know it's always a trade off between value and effort to fix, and in this case, we may decide the amount of effort required to omit the error it is not justified...

**Comment by rofe** [comment]
Another possibility is that ERROR might be the wrong log level for issues related to implementation details like these, and we could consider changing it to WARN or even INFO...

**Comment by koraa** [comment]
I am happy to change it to warn…

**Comment by rofe** [comment]
@trieloff wdyt?
